src/nvim/terminal.c

```diff
@@ -619,6 +619,24 @@ static void buf_set_term_title(buf_T *buf, char *title)
                &err);
 }
 
+// Get the associated buffer for the terminal and return whether it's in a
+// valid state to be used
+static bool term_buf_valid(Terminal *term, buf_T **buf)
+  FUNC_ATTR_NONNULL_ALL
+{
+  *buf = handle_get_buffer(term->buf_handle);
+  bool valid = true;
+  if (!*buf || !(valid = buf_valid(*buf))) {
+    // Destroyed by `close_buffer`. Do not do anything else.
+    if (!valid) {
+      term->buf_handle = 0;
```

Comment:
Do you mean `TerminalJobData`'s `term` member?  `on_job_output` doesn't look at the buffer, which may not exist anymore at the point it's called.